### PR TITLE
Unit tests and Common Args

### PR DESCRIPTION
**Pull request by mikeyhc**

### CODE REVIEW

File: create-session/main.c [hunk 1]
```diff
@@ -26,20 +26,12 @@ main(int argc, char **argv)
...
 	case 'u':
 		if (user_count >= MAX_USERS)
 			eprintf("max users exceeded (%d max)\n", MAX_USERS);
 		users[user_count++] = EARGF(usage());
 		break;
-	case 'h':
-		usage();
-		exit(0);
-		break;
+	COMMON_ARGS;
```
Comment (kouKatsumi) [inline create-session/main.c]: Shouldn't common args be first as can be -h?

Comment (mikeyhc) [reply create-session/main.c]: Order doesn't matter as the arguments are passed in order,  and -h will exit(0) either way

Comment (kouKatsumi) [reply create-session/main.c]: I would like to have evaluation of common ones at the top, thinks it will be more visible.

File: common/arg.h [hunk 2]
```diff
@@ -44,3 +44,16 @@ extern char *argv0;
 				(brk_ = 1, (argv[0][1] != '\0')?\
 					(&argv[0][1]) :\
 					(argc--, argv++, argv[0])))
+
+#define COMMON_ARGS \
+	case 'd':\
+		path = EARGF(usage());\
+		if (!*path) {\
+			eprintf("path was empty\n");\
```
Comment (kouKatsumi) [inline common/arg.h]: Maybe "No path provided"?

File: common/path.c [hunk 3]
```diff
@@ -42,13 +34,15 @@ char*
 build_path(const char *base, const char *suffix)
 {
 	char *path;
-	size_t base_len;
+	size_t base_len, suffix_len;
 
 	base_len = strlen(base);
-	path = malloc(base_len + strlen(suffix) + 1);
+	suffix_len = strlen(suffix);
```
Comment (kouKatsumi) [inline common/path.c:40]: Isn't suffix length fixed?

File: init/init.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+static int create_config(const char *base);
+static void create_db(const char *base);
+void init(const char *path);
+
+/* copy SRC to DEST */
+static int
+cp(const char *src, const char *dest)
+{
+	int in_fd, out_fd;
+	char buf[8192];
```
Comment (kouKatsumi) [inline init/init.c]: Can we use some kind of const here?

File: init/init.c [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	for (tp = TABLE_LIST; *tp != NULL; tp++)
+		db_exec(db, *tp);
+	sqlite3_close(db);
+}
+
+/* create the required directories and files for booker to work */
+void
+init(const char *path)
+{
+	if (exists(path))
```
Comment (kouKatsumi) [inline init/init.c]: {
}

File: init/init.c [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		return 0;
+	}
+	out_fd = open(dest, O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
+	if (out_fd < 0) {
+		fprintf(stderr, "failed to open %s: %s\n", dest,
+				strerror(errno));
+		return 0;
+	}
+
+	while (1) {
```
Comment (kouKatsumi) [inline init/init.c]: Don't like it. Maybe do while? or read in while argument?

File: common/path.c (hunk 3 above)
Comment (mikeyhc) [reply common/path.c:40]: nope, suffix is last part of path

Comment (kouKatsumi) [reply common/path.c:40]: But I see it used only for default value and then it's fixed size?

File: init/init.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+/* create the required directories and files for booker to work */
+void
+init(const char *path)
+{
+	if (exists(path)) {
+		eprintf("%s is already initialized\n", path);
+	}
+	assert(!mkdir(path, 0755));
```
Comment (kouKatsumi) [inline init/init.c]: Would make a variable from permissions for readability.

**Review by kouKatsumi — DISMISSED** [review]

File: common/path.c (hunk 3 above)
Comment (mikeyhc) [reply common/path.c:40]: it is used in config_path, db_path and default_path which have different suffix lengths

**Review by kouKatsumi — APPROVED** [review]

**Pull request merged**